### ISSUE TITLE
Schedule task from another task

### ISSUE THREAD

**Issue by MishaVyb** [taskiq-python/taskiq#163]
Hi, there! First of all, thank you for `taskiq`. I found it really useful and ease to work with. 

And I wonder, is it possible to schedule task from another executable task. For example:

```python
@broker.task('do_something')
async def do_something(arg: int):
    print('do_something...', arg)

@broker.task('tasks_smart_scheduler', schedule=[{"time": datetime.utcnow()}])
async def tasks_smart_scheduler():
    print('tasks_smart_scheduler...')

    # Create do_something task with scheduler label to run it later.
    later = datetime.utcnow() + timedelta(seconds=10)
    decorated_task = broker.task(
        f'do_something:{later}',   # pass custom name to avoid conflicts with initial `do_something` task
        schedule=[
            {
                "time": datetime.utcnow(),
                "args": 123
            }
        ],
    )(do_something)

    assert decorated_task in broker.available_tasks.values()  # Task added to broker successfully
```

- `decorated_task` is not enqueued here, I'm expecting that scheduler will do that on proper time (later). If I wont to enqueue it's wright now, I could write following. But it's not what I'm trying to do.

```python
await decorated_task.kiq(123)
```

So, I run both `worker CLI` and `scheduler CLI`. But, obviously, it does not work, because them running in **separate** processes, and `scheduler` is knowing **nothing** about new scheduled task just created at worker process.

Is it possible to communicate between scheduler and worker? Run them simultaneously in one process somehow? Or it's an anti-pattern at all?

**Comment by s3rius** [comment]
Hi and thanks for raising this issue. Let me answer easy part first.

## Can I run scheduler along with worker?

Yes, you can. Although it is possible to run scheduler along the broker, it is highly not recommended. The main reason is that workers should be able to scale, and scheduler should have only one instance at any given moment. If you're going to have only one worker instance you might start scheduler as a task in your event loop on worker's startup. You can find it's main function [here](https://github.com/taskiq-python/taskiq/blob/00ee6cf66504104b1d22ab7ef5e713511e5c7693/taskiq/cli/scheduler/run.py#L65).

This is an example of how you can do so: 

```python
from taskiq.cli.scheduler.run import run_scheduler
from taskiq.cli.scheduler.args import SchedulerArgs


@broker.on_event(TaskiqEvents.WORKER_STARTUP)
async def my_startup(state: TaskiqState):
    loop = asyncio.get_running_loop()
    state.scheduler_task = loop.create_task(run_scheduler(SchedulerArgs(...)))
```

If you're going to use it that way, please make sure, you have only one instance of the worker at any given moment. Also, this command should not fail, but it might, so you better wrap `run_scheduler` with another async task that restarts the `run_sheduler` if it exists or raises and exception (it should not happen, but anyway). 

## Is it possible to schedule a task from another task.

Sure. But since you would need to have a storage for scheduled tasks, we couldn't come up with generic solution, but we have all the building blocks. 

I would suggest using database to store tasks. We have [ScheduleSource](https://taskiq-python.github.io/extending-taskiq/schedule-sources.html) class which you have to inherit in order to get schedules.

You would have to create a table on startup and in `get_schedules` you would need to do something like this:

```python

class MyDBSource(ScheduleSource):
    ...
    async def get_schedules(self) -> List["ScheduledTask"]:
        data = self.connection.execute("SELECT * FROM myschedules WHERE date>= NOW()")
        return [
            ScheduledTask(
                task_name=row.task_name,
                labels={"schedule_id": row.id, **json.loads(row.labels)},
                args=json.loads(row.args),
                kwargs=json.loads(row.kwargs),
                cron=None,
            ) 
            for row in data
        ]

```

Its' pseudocode and not an actual implementation, but I hope you got the idea. When you scheduler source is ready, 
we can enable your schedulesource. 

To do so, we need to add it to your scheduler.

```python
from taskiq.scheduler import TaskiqScheduler
from taskiq.schedule_sources import LabelScheduleSource

from your_code.my_source import MyDBSource

broker = ...

scheduler = TaskiqScheduler(
    broker=broker,
    sources=[LabelScheduleSource(broker), MyDBSource("psql://meme:meme@github.com:5432/memes")],
)
```

Now you need to create tasks. Since we don't know how and where you're going to store it, you have to write your own function to store it somewhere. Basically it's just an ordinary function and may look something like this:


```python
import datetime
import json
from typing import Any

from taskiq import AsyncTaskiqDecoratedTask


async def save_task(
    db_session: SomeSession,
    task: AsyncTaskiqDecoratedTask,
    schedule: datetime.datetime,
    *args: Any,
    labels: dict[str, Any] = {},
    **kwargs: Any,
):
    await db_session.execute(
        """
        INSERT INTO 
        myschedules (
            task_name, 
            labels, 
            args, 
            kwargs, 
            date
        ) 
        VALUES (
            :task_name, 
            :labels, 
            :args, 
            :kwargs, 
            :date
        )
        """,
        params={
            "task_name": task.task_name,
            "labels": json.dumps(labels),
            "args": json.dumps(args),
            "kwargs": json.dumps(kwargs),
        },
    )

```

And that's it. Now you should be able to create schedules from any of your tasks:

```python
from datetime import datetime, timedelta
from your_project.tkq import broker

@broker.task
async def my_another_task():
    pass


@broker.task
async def my_task(db_session: SomeSession = TaskiqDepends(get_my_session)):
    await save_task(db_session, my_another_task, datetime.utcnow() + timedelta(minutes=10))
```

I hope that I was clear in my explanations and it helped you. Currently I'm updating documentation and I guess I should put this example in common recipes.

**Comment by MishaVyb** [comment]
@s3rius thanks! You was totally clear. I see both ideas are suite my case. I think, I'm gonna extend [ScheduleSource](https://taskiq-python.github.io/extending-taskiq/schedule-sources.html) as you mention above, but with **Redis** as a place to store scheduled tasks (as it more easy to work with).

By the way. While debugging, I've found, that `run_scheduler` kiq task every minute, if it was scheduled with `time` option . If found the implementation of that in this condition:`task.time <= datetime.utcnow()`. When it becomes truly, task `kiq`s every `run_scheduler` loop iteration. 

Is it feature or bug? As for me, task with `time` (not `cron`) option should be enqueued only once.

**Comment by s3rius** [comment]
WOW! Guess it's a bug. We need to fix it, because tasks with time option indeed should be sent once. But I have a question. What broker do you use?

**Comment by MishaVyb** [comment]
Full example below

```python
from datetime import datetime

from taskiq import TaskiqScheduler
from taskiq.schedule_sources import LabelScheduleSource
from taskiq_redis import ListQueueBroker, RedisAsyncResultBackend

redis_async_result = RedisAsyncResultBackend(redis_url="redis://localhost:6379")
broker = ListQueueBroker(url="redis://localhost:6379").with_result_backend(redis_async_result)
scheduler = TaskiqScheduler(
    broker=broker,
    sources=[LabelScheduleSource(broker)],
    refresh_delay=10.0,
)


@broker.task('do_something', schedule=[{"time": datetime.utcnow()}])
async def do_something():
    print('do_something...')
```

**Comment by s3rius** [comment]
Ah, I see now. Guess the problem is that it's not possible to identify whether you want to delete this task or run it every 10 seconds. It's a problem with `LabelScheduleSource`, since it allows only static labeling. It's unfixable and predictable behavior. But I should add this in docs to make them more obvious.

**Comment by s3rius** [comment]
We can have a list of sent tasks, but still, having it all in memory may lead to memory leaks and unpredictable behavior in the future. 

We have this comparison:
```python
    if task.time is not None:
        return task.time <= datetime.utcnow()
```  
because there are some situations where task should have ran in a previous iteration or was added right after the previous iteration of the scheduler. In order to not miss such situations we've had to came up with this comparison. 

Hope it makes sense.

**Comment by MishaVyb** [comment]
> want to run it once or run it every 10 seconds
Not quite. `refresh_delay=10.0` responds for how often updates scheduled tasks list in `schedules_updater`

```python
...
await asyncio.sleep(scheduler.refresh_delay)
```

And it's not quite what I was about. Even if I remove this options (default is 30.0 seconds), it won't resolve that bug, because `while` loop in `run_scheduler` iterates every minute:

```python
        delay = (
            datetime.now().replace(second=1, microsecond=0)
            + timedelta(minutes=1)
            - datetime.now()
        )
        await asyncio.sleep(delay.total_seconds())
```

I see why `It's unfixable and predictable behavior`, but I wonder, what if not storing `list of sent tasks`, but store a copy of `self.broker.available_tasks.items()` and mark each task with `time` option: **is it enqueued or not**. I'm not sure It's good solution, just my thoughts.

**Comment by s3rius** [comment]
:thinking: that actually makes sense. Yes, we need to solve this.

**Comment by MishaVyb** [comment]
Maybe, I'll be not really busy next week and sent a PR later. Is any contribution guide for Taskiq?

**Comment by s3rius** [comment]
Yes. We have a list of rules:
* do not add malware.
* Please make sure your PR actually resolves problem.

That's it. It's better to add tests on things you do, but if you think it's not possible or super hard for you, don't do it. It's better if you're going to use `pre-commit`. But it's okay if you not, since we have lints and style tests in CI.

Thank you very much for your contribution. I'll wait for the PR. If you'll be unable to add this, please write here a message, I'll do it.

**Comment by MishaVyb** [comment]
Since [this](https://github.com/taskiq-python/taskiq/pull/164) PR was merged, I think, this issue may be closed
